api-gateway-config/scripts/lua/management/tenants.lua

```diff
@@ -0,0 +1,269 @@
+    end
+  end
+  -- Error checking
+  local fields = {"namespace", "instance"}
+  for k, v in pairs(fields) do
+    if not decoded[v] then
+      request.err(400, utils.concatStrings({"Missing field '", v, "' in request body."}))
+    end
+  end
+  -- Return tenant object
+  local uuid = existingTenant ~= nil and existingTenant.id or utils.uuid()
+  local tenantObj = {
+    id = uuid,
+    namespace = decoded.namespace,
+    instance = decoded.instance
+  }
+  tenantObj = redis.addTenant(red, uuid, tenantObj)
+  redis.close(red)
+  ngx.header.content_type = "application/json; charset=utf-8"
+  request.success(200, tenantObj)
+end
+
+--- Check for tenant id from uri and use existing tenant if it already exists in redis
+-- @param red Redis client instance
+-- @param uri Uri of request. Eg. /v1/tenants/{id}
+function checkURIForExistingTenant(red, uri)
+  local id, existing
+  local index = 1
+  -- Check if id is in the uri
+  for word in string.gmatch(uri, '([^/]+)') do
+    if index == 3 then
+      id = word
+    end
+    index = index + 1
+  end
+  -- Get object from redis
+  if id ~= nil then
+    existing = redis.getTenant(red, id)
+    if existing == nil then
+      request.err(404, utils.concatStrings({"Unknown Tenant id ", id}))
+    end
+  end
+  return existing
+end
+
+--- Get one or all tenants from the gateway
+-- GET /v1/tenants
+function _M.getTenants()
+  local uri = string.gsub(ngx.var.request_uri, "?.*", "")
+  local queryParams = ngx.req.get_uri_args()
+  local id
+  local index = 1
+  local apiQuery = false
+  for word in string.gmatch(uri, '([^/]+)') do
+    if index == 3 then
+      id = word
+    elseif index == 4 then
+      if word:lower() == 'apis' then
+        apiQuery = true
+      else
+        request.err(400, "Invalid request")
+      end
+    end
+    index = index + 1
+  end
+  if id == nil then
+    getAllTenants(queryParams)
+  else
+    if apiQuery == false then
+      getTenant(id)
+    else
+      getTenantAPIs(id, queryParams)
+    end
+  end
+end
+
+--- Get all tenants in redis
+-- @param queryParams object containing optional query parameters
+function getAllTenants(queryParams)
+  local red = redis.init(REDIS_HOST, REDIS_PORT, REDIS_PASS, 10000)
+  local tenants = redis.getAllTenants(red)
+  redis.close(red)
+  local tenantList
+  if next(queryParams) ~= nil then
+    tenantList = filterTenants(tenants, queryParams);
+  end
+  if tenantList == nil then
+    tenantList = {}
+    for k, v in pairs(tenants) do
+      if k%2 == 0 then
+        tenantList[#tenantList+1] = cjson.decode(v)
+      end
+    end
+    tenantList = cjson.encode(tenantList)
+  end
+  ngx.header.content_type = "application/json; charset=utf-8"
+  request.success(200, tenantList)
+end
+
+--- Filter tenants based on query parameters
+-- @param tenants list of tenants
+-- @param queryParams query parameters to filter tenants
+function filterTenants(tenants, queryParams)
+  local namespace = queryParams['filter[where][namespace]']
```

Comment:
Yes. I implemented it this way because `queryParams` is used elsewhere in the code as well.